Changes in module 'nf-core/bedtools/bamtobed'
--- modules/nf-core/bedtools/bamtobed/main.nf
+++ modules/nf-core/bedtools/bamtobed/main.nf
@@ -1,6 +1,6 @@
 process BEDTOOLS_BAMTOBED {
     tag "$meta.id"
-    label 'process_medium'
+    label 'process_single'
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?

************************************************************
